metadata/metadata.go

```diff
@@ -0,0 +1,123 @@
+package metadata
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"reflect"
+	"time"
+
+	"github.com/mackerelio/mackerel-agent/config"
+	"github.com/mackerelio/mackerel-agent/logging"
+)
+
+var logger = logging.GetLogger("metadata")
+
+// Generator generates metadata
+type Generator struct {
+	Name         string
+	Config       *config.MetadataPlugin
+	Tempfile     string
+	PrevMetadata interface{}
+}
+
+// Fetch invokes the command and returns the result
+func (g *Generator) Fetch() (interface{}, error) {
+	message, stderr, exitCode, err := g.Config.Run()
+
+	if err != nil {
+		logger.Warningf("Error occurred while executing a metadata plugin %q: %s", g.Name, err.Error())
+		return nil, err
+	}
+
+	if stderr != "" {
+		logger.Warningf("metadata plugin %q outputs stderr: %s", g.Name, stderr)
+		// metadata plugin can output message to stderr for debugging and json to stdout
+	}
+
+	if exitCode != 0 {
+		return nil, fmt.Errorf("exits with: %d", exitCode)
+	}
+
+	var metadata interface{}
+	if err := json.Unmarshal([]byte(message), &metadata); err != nil {
+		return nil, fmt.Errorf("outputs invalid JSON: %v", message)
+	}
+
+	return metadata, nil
+}
+
+// Differs returns whether the metadata has been changed or not
+func (g *Generator) Differs(metadata interface{}) bool {
```

Comment:
Hmm, in precise, the (state of) metadata is changed (new metadata differs from previous metadata), and not the generator is changed. Looks oddly a bit for me but I changed to IsChanged https://github.com/mackerelio/mackerel-agent/pull/333/commits/090911319450e92873d25fc639c74dc1c33375b3